projects/components/src/table/table.component.ts

```diff
@@ -407,15 +408,11 @@ export class TableComponent
     this.columnStateSubject.next(sort.column);
   }
 
-  public toggleRowSelection(row: StatefulTableRow): void {
-    row.$$state.selected = !row.$$state.selected;
-
+  public toggleRowSelected(row: StatefulTableRow): void {
     const rowSelections = this.selections ?? [];
-    this.selections = rowSelections.includes(row)
-      ? rowSelections.filter(selection => selection !== row)
-      : rowSelections.concat(row);
-    this.selectionsChange.emit(this.selections);
-
+    const rowIndexInSelections = rowSelections.findIndex(selection => isEqualIgnoreFunctions(selection, row));
+    rowIndexInSelections >= 0 ? rowSelections.splice(rowIndexInSelections, 1) : rowSelections.push(row);
```

Comment:
why do we need to mutate this? This is an external reference that we're mutating, even though we hid it by reassigning it.

Reply:
Either we have to keep a local state to track the state of selections, or mutate the selections reference.

Mutating an Input is not good, because it would indirectly trigger `ngOnChanges`?

Reply:
nope (even if we could/did this outside, onChanges is always based on references). Mutations aren't good because it makes it very difficult to understand the side effects of a change and very bug prone code - this table change feels like an example of that to me. A little more concretely, imagine I'm doing something (outside table) like assigning the array of selected rows to another component to render (say, sankeys). When that reference is mutated, the internals of the table potentially just modified a sibling component (or a parent, in other scenarios). Or maybe the reference was mutated and then emitted through the output - but since the reference is still the same, it doesn't trigger change detection in the sibling.

tl;dr: Angular is built on the idea of unidirectional data flow, and mutation breaks that.

Reply:
I tried maintaining a local state to achieve this, but couldn't make it working

Reply:
ignore the above comment. fixed non mutating the Input selections